### PR TITLE
Light type

### LINKED ISSUE #1032: Add `@gapstack/light-type`
https://github.com/Nick-Lucas/gapstack/blob/main/packages/light-type/README.md

Mentioned here: https://dev.to/nicklucas/typescript-runtime-validators-and-dx-a-type-checking-performance-analysis-of-zodsuperstructyuptypebox-5416

Ping @Nick-Lucas

### PR DESCRIPTION
**Pull request by Nick-Lucas**
Closes #1032

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -24,6 +24,7 @@
   "dependencies": {
     "@ailabs/ts-utils": "1.4.0",
     "@badrap/valita": "0.1.9",
+    "@gapstack/light-type": "^0.0.23",
```
Comment (moltar) [inline package.json:27]:
```suggestion
    "@gapstack/light-type": "0.0.23",
```

File: package-lock.json [hunk 2]
```diff
@@ -11,10 +11,11 @@
       "dependencies": {
         "@ailabs/ts-utils": "1.4.0",
         "@badrap/valita": "0.1.9",
+        "@gapstack/light-type": "^0.0.23",
```
Comment (moltar) [inline package-lock.json:14]:
```suggestion
        "@gapstack/light-type": "0.0.23",
```

**Review by moltar — CHANGES_REQUESTED** [review]

**Comment by moltar** [comment]
Please also add the package to the `README.md` file

**Comment by moltar** [comment]
Lots of this. I think it is due to the older TS version. :/

```
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1738): error TS1005: '?' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1746): error TS1005: ':' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1761): error TS1005: ')' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1763): error TS1068: Unexpected token. A constructor, method, accessor, or property was expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1779): error TS1005: ',' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1783): error TS1005: ';' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1869): error TS1109: Expression expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1877): error TS1434: Unexpected keyword or identifier.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1901): error TS1005: ']' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1902): error TS1005: ',' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1903): error TS1136: Property assignment expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1912): error TS1003: Identifier expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1935): error TS1005: '{' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1986): error TS1005: ',' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1990): error TS1005: ':' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1994): error TS1128: Declaration or statement expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,1996): error TS1434: Unexpected keyword or identifier.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2001): error TS1128: Declaration or statement expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2008): error TS1005: ';' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2059): error TS1005: '(' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2065): error TS1005: ',' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2072): error TS1005: ')' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2074): error TS1128: Declaration or statement expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2076): error TS1128: Declaration or statement expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2078): error TS1434: Unexpected keyword or identifier.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2083): error TS1128: Declaration or statement expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2085): error TS1128: Declaration or statement expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2093): error TS1434: Unexpected keyword or identifier.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2121): error TS1005: ']' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2383): error TS1005: '?' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2391): error TS1005: ':' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2406): error TS1005: ')' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2728): error TS1005: ',' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2737): error TS1005: ':' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2741): error TS1005: ',' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2748): error TS1005: ',' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2749): error TS1136: Property assignment expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2750): error TS1136: Property assignment expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2752): error TS1136: Property assignment expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2837): error TS1109: Expression expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2851): error TS1005: ',' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2869): error TS1005: ']' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2870): error TS1005: ',' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2871): error TS1136: Property assignment expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2880): error TS1003: Identifier expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2903): error TS1005: '{' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2954): error TS1005: ',' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2958): error TS1005: ':' expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2960): error TS1128: Declaration or statement expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2962): error TS1128: Declaration or statement expected.
Error: node_modules/@gapstack/light-type/src/lib/chainable/LtObject.d.ts(26,2964): error TS1434: Unexpected keyword or identifier.
```

**Comment by Nick-Lucas** [comment]
> Lots of this. I think it is due to the older TS version. :/
```

Right I see now this repo is on 4.6, yeah I might be using some newer features.

Is it okay to upgrade typescript in this repo or you're keen to keep it lower for compat reasons?

**Comment by Nick-Lucas** [comment]
Doesn't look like updating TS will be easy in this repo, GTS is already complaining that it only supports up to 4.5.0, while not having a newer release, and running the lint command seems to build the libs and then results in prettier errors for the build folders. Needs some work to replace gts and ignore linting the build dirs.

**Comment by moltar** [comment]
> Is it okay to upgrade typescript in this repo or you're keen to keep it lower for compat reasons?

Yeah, upgrading is easy, just there are some _other_ packages that break then 😁 

Need to decide if we shall be deprecating those packages eventually. @hoeck thoughts?

**Comment by Nick-Lucas** [comment]
In theory I should be able to do some type lowering for light-type,  just need to find the time. 

Though I'm seeing plenty of libraries now requiring 4.7 as a baseline - I guess your own hand might get forced upward in that respect

**Comment by moltar** [comment]
> I guess your own hand might get forced upward in that respect

Yeah, I have nothing against the new features, and in fact, always use latest TS on all my projects.

It is just a decision we have to make, or a complete architecture, which is also probably something would be good, to isolate each package's test environment entirely.

**Comment by hoeck** [comment]
> Yeah, I have nothing against the new features, and in fact, always use latest TS on all my projects.

Same here :+1: 

> Need to decide if we shall be deprecating those packages eventually. @hoeck thoughts?

Keeping lots of different TS versions around and separating them from one another looks like a lot of work to me without much gain.

Performance is not very likely to be affected by different TS versions at all. And even if this were the case, would anyone choose a package because it is a more performant in an *older* version of TS?

> Yeah, upgrading is easy, just there are some other packages that break then

Removing (or disabling) old unmaintained packages seems like a good tradeoff for me.

I'd say lets go check what happens if we bump TS to the latest version :grin:

**Comment by moltar** [comment]
@Nick-Lucas typescript is at `5.1.6` now. Would you like to try again? Thanks.

**Pull request closed without merging**